RFE: cleanup the API return values and document them the in the manpages

pcmoore opened the pull request:
This has already been covered in detail in issue #97 so I won't repeat the details here.  In summary, this PR establishes the APIs error codes as part of the API from here moving forwards.

pcmoore commented:
Forced a push to resolve the merge conflict against master.

drakenclimber commented on src/api.c:
```diff
@@ -610,7 +610,7 @@ API int seccomp_export_bpf(const scmp_filter_ctx ctx, int fd)
 	gen_bpf_release(program);
 	if (rc < 0)
-		return -errno;
+		return -ECANCELED;
```
I'm wondering if we should consider adding a log/debug feature that could capture the `errno` data we're now throwing away.  It may prove help debugging when a customer files a bug against us.

pcmoore replied:
I see your point, I've got an idea, let me see what I can come up with ... stay tuned.

pcmoore replied:
As one potential option, I just added a patch to this PR, "api: add the SCMP_FLTATR_API_SYSRAWRC filter attribute".  I'm going to mark this as resolved and we can continue the conversations in the comments or as review comments in the relevant code.

drakenclimber commented on src/gen_pfc.c:
```diff
@@ -469,18 +469,17 @@ static int _gen_pfc_arch(const struct db_filter_col *col,
 	newfd = dup(fd);
 	if (newfd < 0)
-		return errno;
+		return -ECANCELED;
```
Random low-priority comment - when I was cscoping errno, I noticed some of our functions in util.c and scmp_bpf_disasm.c sometimes return `-errno` and sometimes `errno`.  The doxygen comments of the functions imply they all should be returning a negative error number.

pcmoore replied:
Hmm, good point.  We're not quite extending the same return value promise to these tools (yet), but we should fix that.  I'll add a patch to the next force push.

pcmoore replied:
Now that I'm looking at it, scmp_bpf_disasm.c looks correct, it's just the comments that are wrong.  For the API we want to return negative values on error, but for the terminal apps we want to return a positive errno value.

Does that make sense, or am I crazy?

drakenclimber replied:
Yes, I agree with your analysis.  The use of positive errors on a terminal has always been a pet peeve of mine because it seems so... error prone :(.

Let's fix up the comments some time, but it's obviously not critical or related to this work.

pcmoore replied:
> ... a pet peeve of mine because it seems so... error prone :(

:)

> Let's fix up the comments some time, but it's obviously not critical or related to this work.

I thought I fixed the comments, but I'm sure I missed a few ... can you point me in the right direction?

drakenclimber commented on src/system.c:
```diff
@@ -407,6 +409,6 @@ int sys_notify_id_valid(int fd, uint64_t id)
 
 	if (ioctl(fd, SECCOMP_IOCTL_NOTIF_ID_VALID, &id) < 0)
-		return -errno;
+		return -ENOENT;
```
I didn't expect to see an `-ENOENT` error in this patch, but I think it's the right change to make for this ioctl.

pcmoore replied:
Not sure what to say in response here other than "yup" :)

drakenclimber commented on src/gen_bpf.c:
```diff
@@ -2254,6 +2251,9 @@ struct bpf_program *gen_bpf_generate(const struct db_filter_col *col)
 	struct bpf_program *prgm;
 
+	if (col->filter_cnt == 0)
+		return NULL;
```
Returning NULL here will lead to us returning `-ENOMEM`.  The comment in `_rc_filter()` says this about `-ENOMEM`:

```
/* NOTE: unable to allocate enough memory to perform the
*       requested operation */
```
I am afraid this will cause confusion for some users.  Should we update the above comment?  Or is it fair to return `-EINVAL` because the user tried to build a bpf program with no valid filters in it?

pcmoore replied:
This is tricky as currently `gen_bpf_generate(...)` doesn't really offer a lot of granularity when it comes to return value: it's either a valid pointer or NULL.  If there is any hope, it is that `_gen_bpf_build_bpf(...)` does return an error code and not simply a pointer; we could modify `gen_bpf_generate(...)` to return the `bpf_program` pointer via an argument and return a more useful return code ... assuming we got back and audit all the return codes in "src/gen_bpf.c".

What do you think @drakenclimber?

pcmoore replied:
I just added another patch to this PR which should address this; I'm going to mark this as resolved, if there are any issues let's discuss them against the newly added code.

drakenclimber commented on src/gen_pfc.c:
```diff
@@ -289,7 +289,7 @@ static int _get_bintree_syscall_num(const struct pfc_sys_list *cur,
 
 	if (cur == NULL)
-		return -EINVAL;
+		return -EFAULT;
```
Unrelated to this change - I noticed that `_gen_pfc_arch()`can return a negative return code, but we're ignoring the return code entirely in `gen_pfc_generate()`.  Not a big deal currently, but probably a nitpick worth fixing someday

drakenclimber replied:
Oh, and `-EFAULT` isn't in `_rc_filter()`, but as mentioned in the comment above, these errors are being suppressed entirely.  So not a big deal at the moment.

pcmoore replied:
> I noticed that _gen_pfc_arch()can return a negative return code, but we're ignoring the return code entirely in gen_pfc_generate(). Not a big deal currently, but probably a nitpick worth fixing someday.

Seems related to the discussion above, I'm going to mark this comment resolved in favor of discussion it up there.

> Oh, and -EFAULT isn't in _rc_filter() ...

Well, since the default is `-EFAULT` it doesn't need to be :)

>

drakenclimber commented on src/api.c:
```diff
@@ -210,7 +255,7 @@ API int seccomp_reset(scmp_filter_ctx ctx, uint32_t def_action)
 	if (ctx == NULL || db_col_action_valid(NULL, def_action) < 0)
```
Argh - github won't let me comment on code outside of the area that was changed.

Do you need a `_rc_filter()` call on line 230 of src/api.c in `seccomp_api_set()`?  It's returning `-EINVAL` without being checked

pcmoore replied:
Ha, `seccomp_api_set()` was the other function I caught went I double checked src/api.c :)

Thanks for noticing this, should be fixed in the next force push.

drakenclimber commented on src/api.c:
```diff
@@ -210,7 +255,7 @@ API int seccomp_reset(scmp_filter_ctx ctx, uint32_t def_action)
 	if (ctx == NULL || db_col_action_valid(NULL, def_action) < 0)
 		return -EINVAL;
```
Do you want to call `_rc_filter()` here as well?

pcmoore replied:
Good catch, thanks.  I did a second check and caught a few others in src/api.c.  This should be fixed in the next force push.

drakenclimber commented on src/api.c:
```diff
@@ -653,9 +653,9 @@ API int seccomp_export_bpf(const scmp_filter_ctx ctx, int fd)
-	program = gen_bpf_generate((struct db_filter_col *)ctx);
-	if (program == NULL)
-		return _rc_filter(-ENOMEM);
+	rc = gen_bpf_generate((struct db_filter_col *)ctx, &program);
```
Thanks for making this change.  I think this will really help in debugging

pcmoore commented:
As mentioned in the review, I added a new patch which introduces a new filter attribute, `SCMP_FLTATR_API_SYSRAWRC` which I think should allow callers to chose between a stable set of return codes, and the full set of return codes provided by the system.  Unfortunately I wish we didn't have to choose, but given the wide range of kernels, libcs, etc. I'm not sure how we do that.  If anyone has any ideas I would *love* to hear them.

I will admit to somewhat rushing the patches today as I wanted to get them out for review; I'll give them a better look tomorrow, but I wanted to see if the filter attribute idea seemed reasonable.

pcmoore commented:
> I briefly looked through the RAWRC patch. It seems like a solid solution to me although I would like to think about it a bit more.

I understand, it is far from perfect but I'm struggling to come up with something better.  I'm always open to ideas if anyone has any.  @drakenclimber I'm going to hold off on adding your ACK and pushing until I hear an "I can't think of anything better, I'm okay with this".

I also just did a force push which adds some tests and rebases against our current master branch.

pcmoore commented:
FYI, for some reason Travis canceled a could of the PR runs around this time so something must of been wonky with the infrastructure - I think we can ignore the current failed Travis run for now.

drakenclimber commented:
> I understand, it is far from perfect but I'm struggling to come up with something better. I'm always open to ideas if anyone has any. @drakenclimber I'm going to hold off on adding your ACK and pushing until I hear an "I can't think of anything better, I'm okay with this".

Same here.  I think this is a good first step until we think of something better.  I'm good with the RAWRC changes.

pcmoore commented:
Thanks @drakenclimber, I think we're on the same page here.  Although one thing to keep in mind is that API changes such as the filter attribute are forever - we don't get to back them out.  That said, it's a simple enough API that it should give us some wiggle room if needed.

I've rebased this PR on top of master, added your ACKs, and force pushed.  Assuming Travis runs clean I'll go ahead and merge the PR.  If you have any second thoughts we can always revert the changes before we release v2.5.0.
